File: apps/api/app/Models/MarketQuote.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class MarketQuote extends Model
{
    use HasFactory;

    protected $fillable = [
        'security_id',
        'quote_time',
        'last_price',
        'pre_close',
        'open',
        'high',
        'low',
        'volume',
        'amount',
        'turnover_rate',
        'pct_change',
        'provider',
        'source_timestamp',
        'metadata',
    ];

    protected $casts = [
        'quote_time' => 'datetime',
        'source_timestamp' => 'datetime',
        'metadata' => 'array',
        'last_price' => 'decimal:4',
        'pre_close' => 'decimal:4',
        'open' => 'decimal:4',
        'high' => 'decimal:4',
        'low' => 'decimal:4',
        'volume' => 'decimal:4',
        'amount' => 'decimal:4',
        'turnover_rate' => 'decimal:6',
        'pct_change' => 'decimal:6',
    ];

    public function security(): BelongsTo
    {
        return $this->belongsTo(Security::class);
    }
}

File: services/intelligence/app/models/market_quote.py
from sqlalchemy import JSON, DateTime, ForeignKey, Numeric, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class MarketQuote(Base):
    __tablename__ = "market_quotes"
    __table_args__ = (UniqueConstraint("security_id", "quote_time", name="uk_security_quote_time"),)

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    security_id: Mapped[int] = mapped_column(ForeignKey("securities.id"), nullable=False)
    quote_time: Mapped[DateTime] = mapped_column(DateTime, nullable=False)
    last_price: Mapped[float | None] = mapped_column(Numeric(18, 4), nullable=True)
    pre_close: Mapped[float | None] = mapped_column(Numeric(18, 4), nullable=True)
    open: Mapped[float | None] = mapped_column(Numeric(18, 4), nullable=True)
    high: Mapped[float | None] = mapped_column(Numeric(18, 4), nullable=True)
    low: Mapped[float | None] = mapped_column(Numeric(18, 4), nullable=True)
    volume: Mapped[float | None] = mapped_column(Numeric(24, 4), nullable=True)
    amount: Mapped[float | None] = mapped_column(Numeric(24, 4), nullable=True)
    turnover_rate: Mapped[float | None] = mapped_column(Numeric(12, 6), nullable=True)
    pct_change: Mapped[float | None] = mapped_column(Numeric(12, 6), nullable=True)
    provider: Mapped[str] = mapped_column(String(32), nullable=False)
    source_timestamp: Mapped[DateTime | None] = mapped_column(DateTime, nullable=True)
    metadata_json: Mapped[dict | None] = mapped_column("metadata", JSON, nullable=True)
    created_at: Mapped[DateTime | None] = mapped_column(DateTime, nullable=True)
    updated_at: Mapped[DateTime | None] = mapped_column(DateTime, nullable=True)
